sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala

```diff
@@ -118,13 +119,28 @@ private[sql] object JDBCRelation extends Logging {
           s"Upper bound: ${boundValueToString(upperBound)}.")
         upperBound - lowerBound
       }
-    // Overflow and silliness can happen if you subtract then divide.
-    // Here we get a little roundoff, but that's (hopefully) OK.
-    val stride: Long = upperBound / numPartitions - lowerBound / numPartitions
+
+    // Overflow can happen if you subtract then divide. For example:
+    // (Long.MaxValue - Long.MinValue) / (numPartitions - 2).
+    // Also, using fixed-point decimals here to avoid possible inaccuracy from floating point.
+    val strideUpperCalculation = (upperBound / BigDecimal(numPartitions))
+      .setScale(18, RoundingMode.HALF_EVEN)
+    val strideLowerCalculation = (lowerBound / BigDecimal(numPartitions))
+      .setScale(18, RoundingMode.HALF_EVEN)
```

Comment:
Why do we use `scale=18` and `RoundingMode.HALF_EVEN` here?

Reply:
I'm using setScale here to handle an issue that can arise with repeating decimals. For instance, if you end up with upperStride = 1.333... and lowerStride of  0.333..., when doing upperStride - lowerStride, you'd get 0.999.... This would then be truncated to 0 when converted to a long for stride, which is an issue. Using setScale, the repeating decimals are rounded and after the subtraction you end up with 1.0. 

The reason I'm using `RoundingMode.HALF_EVEN` is because that's the default for BigDecimal. As for a scale of 18, this was a number I picked from prior defaults I've encountered (e.g. when a DecimalType is deserialized into BigDecimal, the BigDecimal is instantiated as 38,18). However, I could lower the scale if you think it is worthwhile. Since these calculations are at such a small scale, I wasn't concerned about BigDecimal performance, but instead accuracy.

Reply:
I'll go ahead and mark this as resolved for now. We can unresolve it if you, or someone else, have a different suggestion.